### PR TITLE
Pierce/reduction fps_bitrate

### PR DESCRIPTION
**Pull request by PierceLovesee**
Reduction method using change_fps and set_bitrate in aEye.Labeler.

Input video taken from: S3:aeye-data-bucket/input_video/
Outputs video files to: S3:leto-dish/reduction/fps_bitrate/

Running instructions in the README.

### CODE REVIEW

File: leto/reduction/fps_bitrate/requirements.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+aEye>=1.1.0
```
Comment (JamesF2440) [inline leto/reduction/fps_bitrate/requirements.txt:1]: nice 😎

File: leto/reduction/fps_bitrate/fps_bitrate.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    Returns
+    ----------
+        video_list: list
+            list of videos with changes labeled for bitrate and framerate reduction; to be processed in runner method
+    """
+
+    labeler = Labeler()
+
+    try:
+        if fps <= 0:
```
Comment (JamesF2440) [inline leto/reduction/fps_bitrate/fps_bitrate.py:39]: Should just be < 1, as 0 FPS makes FFmpeg freak out a bit

File: leto/reduction/app.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    logging.info('succesfully loaded function')
+
+    aux=Aux()
+    video_list = aux.load_s3(bucket = 'aeye-data-bucket', prefix = 'input_video/')
+    reduction(video_list)
+    out = aux.execute_label_and_write_local(video_list)
+    aux.upload_s3(video_list, "leto-dish", "reduction/")
+    aux.clean()
+
+    return 'video reduction completed on ' + sys.version + '!'
```
Comment (mg6955) [inline leto/reduction/app.py]: Why does the function need to return? Can you make it into a logging statement?

File: leto/reduction/fps_bitrate/lambda_function.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    except Exception as e:
+        print(e)
+        logging.warning(f"unable to load video list from s3.")
+    
+    fps_bitrate(video_list)
+    out = aux.execute_label_and_write_local(video_list)
+    aux.upload_s3(video_list, "leto-dish", "reduction/")
+    aux.clean()
+
+    return 'video reduction completed on ' + sys.version + '!'
```
Comment (mg6955) [inline leto/reduction/fps_bitrate/lambda_function.py]: Can you make it into a logging statement?

File: leto/reduction/fps_bitrate/app_fps_bitrate.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    except Exception as e:
+        print(e)
+        logging.warning(f"unable to load video list from s3; ensure AWS credentials have been provided.")
+    
+    fps_bitrate(video_list)
+    out = aux.execute_label_and_write_local(video_list)
+    aux.upload_s3(video_list, "leto-dish", "reduction/fps_bitrate/")
+    aux.clean()
+
+    return 'video reduction completed on ' + sys.version + '!'
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/app_fps_bitrate.py]: got it changed 👍

File: leto/__init__.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#from .downstream_model import ...
+
+
+#from .reconstruction import ...
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/__init__.py]: Removed this file to shift scope of PR entirely to getting the reduction module, fps_bitrate, merged in.  Restructure of the repo later.

File: leto/reduction/fps_bitrate/README.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# Reduction using fps_bitrate
+
+1. Provide AWS credentials
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/README.md]: Added some additional verbiage surrounding this.  Take a look and let me know what you think.

File: leto/reduction/fps_bitrate/fps_bitrate.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
+from aEye import Labeler 
+import logging
+
+def fps_bitrate(video_list, fps=30, bitrate=0):
+
+    """
+    Wrapper method for the change_fps and set_bitrate methods in aEye.Labeler.
+
+    This method addes labels to the input videos for fps and bitrate rectuion.
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/fps_bitrate.py]: 👍

File: leto/reduction/fps_bitrate/fps_bitrate.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    The lables are then exectured in the runner method to process the videos.
+    Default settings of fps=30 and bitrate=0 result in the output videos having
+    30 frames per second and a 10x reduction in bitrate.
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/fps_bitrate.py]: Good catch!  Got that more explicit for the file bitrate.

File: leto/reduction/fps_bitrate/README.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+2. Ensure you are in the ***fps_bitrate*** directory:
+
+2. Install requierments:
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/README.md]: Good catch 👍

File: leto/reduction/fps_bitrate/README.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+```console
+pip install -r requirements.txt
+```
+
+3. execute the runner method, ***app_fps_bitrate.py***
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/README.md]: Love the nit and the pick

File: leto/reduction/fps_bitrate/app_fps_bitrate.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def app_fps_bitrate():
+
+    """
+    Runner method for fps_bitrate.fps_bitrate().  This method abstracts some of the 
+    interaction with S3 and AWS away from fps_bitrate.
+
+    Input and Output S3 paths are hard coded in this version; this can be abstracted in 
+    future versions to allow for dynamic Input and Output allocation.
```
[the remark this answers is not in the document]
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/app_fps_bitrate.py]: I appreciate the feedback -- I would like to leave it as is for the time being.  If we have those S3 paths locked down, it'll make debugging the end-to-end pipeline easier.  Then, if we move forward with this reduction technique, making it dynamic is a 5 minute PR.

File: leto/reduction/fps_bitrate/requirements.txt (hunk 1 above)
Comment (PierceLovesee) [reply leto/reduction/fps_bitrate/requirements.txt]: Love to see it

File: leto/reduction/app.py (hunk 3 above)
Comment (PierceLovesee) [reply leto/reduction/app.py]: Got it changed to a logging.info statement 🙏

**Review by mg6955 — APPROVED** [review]

**Pull request merged**